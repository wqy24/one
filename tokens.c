#include "tokens.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdint.h>

#define EXPAND_RATE 32

int pred_ref(char c) { return c != '}'; }
int pred_str(char c) { return c != '`'; }
int pred_num(char c) { return (isdigit(c) || c == '.'); }
int pred_idn(char c) { return (isalnum(c) || c == '_' || c == '-'); }

char *read_value(char first, FILE *input, int (*pred)(char), int rm_quote)
{
  size_t size = EXPAND_RATE * sizeof(char);
  char *buffer = (char *)malloc(size), *res;
  char c;
  size_t count = 0;
  if (!rm_quote) buffer[count++] = first;

  while ((c = getc(input)) != EOF && pred(c)) {
    buffer[count++] = c;
    if (count + 1 >= size) {
      size += EXPAND_RATE * sizeof(char);
      char *tmp = realloc(buffer, size);
      if (tmp == NULL) {
        fputs("Out of memory\n", stderr);
        free(buffer);
        return NULL;
      }
      buffer = tmp;
    }
  }

  if (rm_quote && pred(c)) {
    fputs("EOF while parsing a quoted token\n", stderr);
    free(buffer);
    return NULL;
  }

  buffer[count] = '\0';
  res = realloc(buffer, (count + 1) * sizeof(char));
  if (res == NULL) {
    fputs("Out of memory\n", stderr);
    free(buffer);
    return NULL;
  }

  if (!rm_quote && !pred(c)) {
    fseek(input, -1, SEEK_CUR);
  }

  return res;
}

one_token read_one_token(FILE *input)
{
  char curr;

  switch (curr = getc(input)) {
    case EOF: return (one_token){FEOF, NULL};
    case '@': return (one_token){AT, NULL};
    case '^': return (one_token){CARET, NULL};
    case '&': return (one_token){ET, NULL};
    case '(': return (one_token){L_PRN, NULL};
    case ')': return (one_token){R_PRN, NULL};
    case '[': return (one_token){L_BRK, NULL};
    case ']': return (one_token){R_BRK, NULL};
    case '|': return (one_token){PIPE, NULL};
    case ':': return (one_token){COLON, NULL};
    case '<': return (one_token){LT, NULL};
    case '>': return (one_token){GT, NULL};
    case '?': return (one_token){QUESTION, NULL};
    case '~': return (one_token){WAVE, NULL};
    case ';':
      while(!feof(input) && getc(input) != '\n') /* do nothing */;
      return read_one_token(input);
    case ',': return (one_token){COMMA, NULL};
    case '{': return (one_token){REF, read_value('{', input, pred_ref, 1)};
    case '`': return (one_token){STR, read_value('`', input, pred_str, 1)};
    default:
      if (isspace(curr)) return read_one_token(input);
      if (isdigit(curr) || curr == '-') return (one_token){NUM, read_value(curr, input, pred_num, 0)};
      if (isalpha(curr) || curr == '_') return (one_token){IDN, read_value(curr, input, pred_idn, 0)};
      fprintf(stderr, "Invalid character %c.\n", curr);
      exit(EXIT_FAILURE);
  }
}
