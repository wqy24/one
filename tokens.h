#ifndef __ONE_TOKENS_H__
#define __ONE_TOKENS_H__

#include <stdio.h>

typedef struct token {
  enum {
    FEOF,
    STR,
    NUM,
    REF,
    IDN,
    L_BRK,
    R_BRK,
    L_PRN,
    R_PRN,
    PIPE,
    CARET,
    COLON,
    COMMA,
    LT, GT,
    ET /* & */, // AT /* @ */,
    QUESTION,
    WAVE
  } type;
  char *value;
} one_token;

one_token read_one_token(FILE *input); /* malloc warning: "value" in the token */

#endif
