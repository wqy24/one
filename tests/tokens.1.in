; Hello World

`Hello, world`

; map for arr
?a: ?b: proj^a~b: arr^num~a: index^num: proj (arr index) 

; Area of triangle
a^num: b^num: c^num: 
  {num/add} | add^auto:
  {num/sub} | sub^auto:
  {num/mul} | mul^auto:
  {num/div} | div^auto:
  {num/sqrt}| sqrt^auto:
  div (a | add b | add c) 2 
    | p^num:
  sqrt (p | mul (sub a p) | mul (sub b p) | mul (sub c p))

; take
limit^num: ?a: array^num~a: index^num:
  ()
  | for ({num/lt} limit num) (array index)

; drop
drop-items^num: ?a: array^num~a: index^num:
  ()
  | for ({num/gte} 0 num) (array ({num/sub} drop-items index))

` fdafahwlnva 
	a ldkfasd 32823423$52o42934nm, na0-  ==='; a pOIJA I[ jd ndi jh vndhf skjs ad fwe fh;og'glg;j j'';jj'j'"""Jd }{`
