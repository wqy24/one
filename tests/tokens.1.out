STR
Hello, world
QUESTION
IDN
a
COLON
QUESTION
IDN
b
COLON
IDN
proj
CARET
IDN
a
WAVE
IDN
b
COLON
IDN
arr
CARET
IDN
num
WAVE
IDN
a
COLON
IDN
index
CARET
IDN
num
COLON
IDN
proj
L_PRN
IDN
arr
IDN
index
R_PRN
IDN
a
CARET
IDN
num
COLON
IDN
b
CARET
IDN
num
COLON
IDN
c
CARET
IDN
num
COLON
REF
num/add
PIPE
IDN
add
CARET
IDN
auto
COLON
REF
num/sub
PIPE
IDN
sub
CARET
IDN
auto
COLON
REF
num/mul
PIPE
IDN
mul
CARET
IDN
auto
COLON
REF
num/div
PIPE
IDN
div
CARET
IDN
auto
COLON
REF
num/sqrt
PIPE
IDN
sqrt
CARET
IDN
auto
COLON
IDN
div
L_PRN
IDN
a
PIPE
IDN
add
IDN
b
PIPE
IDN
add
IDN
c
R_PRN
NUM
2
PIPE
IDN
p
CARET
IDN
num
COLON
IDN
sqrt
L_PRN
IDN
p
PIPE
IDN
mul
L_PRN
IDN
sub
IDN
a
IDN
p
R_PRN
PIPE
IDN
mul
L_PRN
IDN
sub
IDN
b
IDN
p
R_PRN
PIPE
IDN
mul
L_PRN
IDN
sub
IDN
c
IDN
p
R_PRN
R_PRN
IDN
limit
CARET
IDN
num
COLON
QUESTION
IDN
a
COLON
IDN
array
CARET
IDN
num
WAVE
IDN
a
COLON
IDN
index
CARET
IDN
num
COLON
L_PRN
R_PRN
PIPE
IDN
for
L_PRN
REF
num/lt
IDN
limit
IDN
num
R_PRN
L_PRN
IDN
array
IDN
index
R_PRN
IDN
drop-items
CARET
IDN
num
COLON
QUESTION
IDN
a
COLON
IDN
array
CARET
IDN
num
WAVE
IDN
a
COLON
IDN
index
CARET
IDN
num
COLON
L_PRN
R_PRN
PIPE
IDN
for
L_PRN
REF
num/gte
NUM
0
IDN
num
R_PRN
L_PRN
IDN
array
L_PRN
REF
num/sub
IDN
drop-items
IDN
index
R_PRN
R_PRN
STR
 fdafahwlnva 
	a ldkfasd 32823423$52o42934nm, na0-  ==='; a pOIJA I[ jd ndi jh vndhf skjs ad fwe fh;og'glg;j j'';jj'j'"""Jd }{
FEOF
